src/Microsoft.Identity.Client/Platforms/iOS/iOSBroker.cs

```diff
@@ -0,0 +1,224 @@
+using Microsoft.Identity.Client.Exceptions;
+
+namespace Microsoft.Identity.Client.Platforms.iOS
+{
+    /// <summary>
+    /// Handles requests which invoke the broker. This is only for mobile (iOS and Android) scenarios.
+    /// </summary>
+    internal class iOSBroker : IBroker
+    {
+        private static SemaphoreSlim brokerResponseReady = null;
+
+        private static NSUrl brokerResponse = null;
+        private readonly ApplicationConfiguration brokerRequest;
+        private Dictionary<string, string> _brokerPayload;
+
+        private readonly ICoreLogger _logger;
+        private IServiceBundle _serviceBundle;
+
+        public iOSBroker(ICoreLogger logger)
+        {
+            _logger = logger;
+        }
+
+        public bool CanInvokeBroker
+        {
+            get
+            {
+                CoreUIParent uiParent = new CoreUIParent();
+                if (!brokerRequest.IsBrokerEnabled)
+                {
+                    return false;
+                }
+
+                var result = false;
+
+                if (brokerRequest.IsBrokerEnabled)
+                {
+                    uiParent.CallerViewController.InvokeOnMainThread(() =>
+                    {
+                        result = UIApplication.SharedApplication.CanOpenUrl(new NSUrl("msauthv2://"));
+                    });
+                }
+
+                return result;
+            }
+        }
+
+        public async Task<MsalTokenResponse> AcquireTokenUsingBrokerAsync(Dictionary<string, string> brokerPayload, IServiceBundle serviceBundle)
+        {
+            _serviceBundle = serviceBundle;
+            _brokerPayload = brokerPayload;
+
+            CheckBrokerPayloadForSilentFlow();
+
+            AddIosSpecificParametersToPayload();
+
+            brokerResponse = null;
+            brokerResponseReady = new SemaphoreSlim(0);
+
+            InvokeIosBroker();
+
+            await brokerResponseReady.WaitAsync().ConfigureAwait(false);
+
+            return ProcessBrokerResponse();
+        }
+
+        private void CheckBrokerPayloadForSilentFlow()
+        {
+            if (_brokerPayload.ContainsKey(BrokerParameter.SilentBrokerFlow))
+            {
+                throw new MsalUiRequiredException(MsalError.FailedToAcquireTokenSilently, MsalErrorMessage.FailedToAcquireTokenSilently);
+            }
+        }
+
+        private void AddIosSpecificParametersToPayload()
+        {
+            string base64EncodedString = Base64UrlHelpers.Encode(BrokerKeyHelper.GetRawBrokerKey(_logger));
+            _brokerPayload[iOSBrokerConstants.BrokerKey] = base64EncodedString;
+            _brokerPayload[iOSBrokerConstants.MsgProtocolVer] = "3";
+
+            if (_brokerPayload.ContainsKey(iOSBrokerConstants.Claims))
+            {
+                _brokerPayload.Add(iOSBrokerConstants.SkipCache, "YES");
+                string claims = Base64UrlHelpers.Encode(_brokerPayload[BrokerParameter.Claims]); //TODO: check this
```

Comment:
This is somewhat bad design that you have to overwrite the claims param. But I am sure that after the refactorings suggested this will just go away.